common/leaderelection/leaderelection.go

```diff
@@ -179,3 +188,67 @@ func (e *natsEventBusElector) RunOrDie(ctx context.Context, callbacks LeaderCall
 		}
 	}
 }
+
+type kubernetesElector struct {
+	namespace string
+	leasename string
+	hostname  string
+}
+
+func newKubernetesElector(namespace string, leasename string, hostname string) (Elector, error) {
+	return &kubernetesElector{
+		namespace: namespace,
+		leasename: leasename,
+		hostname:  hostname,
+	}, nil
+}
+
+func (e *kubernetesElector) RunOrDie(ctx context.Context, callbacks LeaderCallbacks) {
+	logger := logging.FromContext(ctx)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		logger.Fatalw("Failed to retrieve kubernetes config", zap.Error(err))
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		logger.Fatalw("Failed to create kubernetes client", zap.Error(err))
+	}
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      e.leasename,
+			Namespace: e.namespace,
+		},
+		Client: client.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: e.hostname,
+		},
+	}
+
+	for {
```

Comment:
There's no need to to have a `for/select` here, it's not working as expected - when it executing `default`, where `RnOrDie` is blocking, even though there's SIGTERM, the `case <-ctx.Done()` won't be executed.

Simply removing the `for/select` should work.

Reply:
That was my initial inclination too, but it turns out (oddly imo) that `leaderelection.RunOrDie` will stop blocking if a pod is the leader and then stops being the leader. The for loop ensures the pod puts itself back in the leader election pool and the `<-ctx.Done()` case ensures that the function will exit gracefully if there's a `SIGTERM`.

Reply:
When a sigterm occurs the parent context is cancelled, which causes all child child contexts to be cancelled, including the one passed to `leaderelection.RunOrDie`. This causes the leader election to halt and for the first branch of the select statement to be invoked on the next iteration of the for loop. I can confirm this behaviour when running locally.